src/uart.c

```diff
@@ -0,0 +1,89 @@
+#include "uart.h"
+#include "utils.h"
+#include "gpio.h"
+
+/* CR1 register bit offsets */
+#define UE_OFFSET       (0)
+#define RE_OFFSET       (2)
+#define TE_OFFSET       (3)
+#define M0_OFFSET       (12)
+#define M1_OFFSET       (28)
+
+/* CR2 register bit offsets */
+#define STOP_OFFSET     (12)
+
+/* ISR register bit offsets */
+#define RNXE_OFFSET     (5)
+#define TC_OFFSET       (6)
+#define TXE_OFFSET      (7)
+
+void    uart_configure(UART_t *uart, UART_config *config)
+{
+    // TODO: add all UART ports
+    if (uart == UART1)
+    {
+        gpio_select_alternate_function(GPIOA, GPIO_PIN_8, AF1);
+        gpio_select_alternate_function(GPIOA, GPIO_PIN_9, AF1);
+        gpio_select_alternate_function(GPIOA, GPIO_PIN_10, AF1);
+    }
+    else if (uart == UART2)
+    {
+        gpio_select_alternate_function(GPIOA, GPIO_PIN_2, AF1);
+        gpio_select_alternate_function(GPIOA, GPIO_PIN_3, AF1);
+        gpio_select_alternate_function(GPIOA, GPIO_PIN_4, AF1);
+    }
+
+    /* Disable UART */
+    uart->CR1 = set_bits_with_offset(uart->CR1, UE_OFFSET, 1, 0);
+
+    switch (config->word_length)
+    {
+        case EIGHT_BITS:
+            uart->CR1 = set_bits_with_offset(uart->CR1, M1_OFFSET, 1, 0);
+            uart->CR1 = set_bits_with_offset(uart->CR1, M0_OFFSET, 1, 0);
```

Comment:
Avoid writing to volatile registers twice. The compiler is not allowed to optimize these into one read and one write.

```c
uart->CR1 = set_bits_with_offset(set_bits_with_offset(uart->CR1, M1_OFFSET, 1, 0), M0_OFFSET, 1, 0);
```